vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,647 @@
+/************************************************************************
+ * This file introduces a class to model a Routing Resource Graph (RRGraph or RRG) 
+ * which is widely used by placers, routers, analyzers etc.
+ *
+ * Overview
+ * ========
+ * RRGraph aims to describe in a general way how routing resources are connected
+ * in a FPGA fabric.
+ * It includes device-level information for routing resources, 
+ * such as the physical location of nodes, 
+ * routing segment information of each routing resource 
+ * switch information of each routing resource.
+ * 
+ * A Routing Resource Graph (RRGraph or RRG) is a Directed Acyclic Graph (DAG),
+ * which consists of a number of nodes and edges.
+ *
+ * Node
+ * ----
+ * Each node represents a routing resource, which could be 
+ * 1. a routing track in X-direction or Y-direction (CHANX or CHANY) 
+ * 2. an input or an output of a logic block (IPIN or OPIN) 
+ * 3. a virtual source or sink node (SOURCE or SINK), which are starting/ending points of routing trees.
+ *
+ * Edge
+ * ----
+ * Each edge represents a switch between routing resources, which could be
+ * 1. a multiplexer
+ * 2. a tri-state buffer
+ * The switch information are categorized in the rr_switch_inf of RRGraph class.
+ *
+ * Guidlines on using the RRGraph data structure 
+ * =============================================
+ *
+ * For those want to access data from RRGraph
+ * ------------------------------------------
+ * Please refer to the detailed comments on each public accessors
+ *
+ * For those want to build/modify a RRGraph
+ * -----------------------------------------
+ * Do NOT add a builder to this data structure!
+ * Builders should be kept as free functions that use the public mutators
+ * We suggest developers to create builders in separated C/C++ source files
+ * outside the rr_graph header and source files
+ *
+ * After build/modify a RRGraph, please do run a fundamental check, a public accessor.
+ * to ensure that your RRGraph does not include invalid nodes/edges/switches/segements 
+ * as well as connections.
+ * This is a must-do check! 
+ *
+ * Example: 
+ *    RRGraph rr_graph;
+ *    ... // Building RRGraph
+ *    rr_graph.check(); 
+ *
+ * Optionally, we strongly recommend developers to run an advance check in check_rr_graph()  
+ * This guarantees legal and routable RRGraph for VPR routers.
+ *
+ * Note that these are generic checking codes. 
+ * If your RRGraph is designed for a special application context, please develop your 
+ * check_rr_graph() function to guarantee its correctness!
```

Comment:
Replace with:

This checks for connectivity or other information in the rr-graph that is unexpected or unusual in an FPGA, and likely indicates a problem in your graph generation. However, if you are intentionally creating an rr-graph with this unusual, buts still technically legal, behaviour, you can write your own check_rr_graph() with weaker assumptions.